tests/end_to_end/server.py

```diff
@@ -1,146 +1,202 @@
+            assert False, failMessage
+        return not self.isRunning()
 
-    def getServerCmdLine(self):
-        self.serverCmdLine = """
-python server_dev.py
---dont-use-reloader
---config TestConfig
---port {} """.format(self.port)
-        return self.serverCmdLine
+    @utils.Timeout()
+    @utils.Repeat()
+    def _waitForErrLines(self):
+        # not sure why there's some delay in getting the server
+        # process' stderr (at least for the ga4gh server)...
+        return self.getErrLines() == []
+
+    def _assertServerShutdown(self):
+        shutdownString = "Server did not shut down correctly"
+        assert self.server.returncode is not None, shutdownString
+        assert not self.isRunning(), shutdownString
 
 
-class ServerTestConfigFile(ServerTest):
+class Ga4ghServerForTesting(ServerForTesting):
     """
-    Launch a server test, but with a custom configuration file
+    A ga4gh test server
     """
-    def run(self, *args, **kwargs):
-        self.config = None
+    def __init__(self):
+        super(Ga4ghServerForTesting, self).__init__(ga4ghPort)
         self.configFile = None
-        self.configFilePath = None
-        super(ServerTestConfigFile, self).run(*args, **kwargs)
-
-    def setConfig(self):
-        raise NotImplementedError("ServerTestConfigFile must be subclassed")
 
-    def getServerCmdLine(self):
-        self.setConfig()
-        self.configFile = tempfile.NamedTemporaryFile()
-        self.configFile.write(self.config)
+    def getConfig(self):
+        config = """
+SIMULATED_BACKEND_NUM_VARIANT_SETS = 10
+SIMULATED_BACKEND_VARIANT_DENSITY = 1
+DATA_SOURCE = "__SIMULATED__"
+DEBUG = True"""
+        return config
+
+    def getCmdLine(self):
+        if self.configFile is None:
+            self.configFile = tempfile.NamedTemporaryFile()
+        config = self.getConfig()
+        self.configFile.write(config)
         self.configFile.flush()
-        self.configFilePath = self.configFile.name
-        self.serverCmdLine = """
+        configFilePath = self.configFile.name
+        cmdLine = """
 python server_dev.py
 --dont-use-reloader
 --config TestConfig
 --config-file {}
---port {} """.format(self.configFilePath, self.port)
-        return self.serverCmdLine
+--port {} """.format(configFilePath, self.port)
+        return cmdLine
+
+    def shutdown(self):
+        super(Ga4ghServerForTesting, self).shutdown()
+        if self.configFile is not None:
+            self.configFile.close()
+
+    def printDebugInfo(self):
+        super(Ga4ghServerForTesting, self).printDebugInfo()
+        print('*** Server CONFIG ***')
+        print(self.getConfig())
+
+
+class TwistedServerForTesting(ServerForTesting):
+    """
+    A twistd test server
+    """
+    def __init__(self, path):
+        super(TwistedServerForTesting, self).__init__(twistdPort)
+        self.path = path
+
+    def getCmdLine(self):
+        cmdLine = "twistd -no web -p {} --path={}".format(
```

Comment:
I think it would be better if we created a Twisted HTTP server in code, and ran this in a subprocess. We're implicitly depending on Twisted here anyway (which we'll need to add to requirements.txt), and it would be good to have direct access to the HTTP server we're using. You should be able to make a HTTP server using twisted in a couple of lines of code.

So, to be clear, I think we should make a function that runs a twisted HTTP server and run this using multiprocessing, rather than calling the shell command `twistd`.